File: dart/glue/lib/src/lib/math/utility/utility.dart
import 'package:glue/src/ir.dart';
import 'package:glue/src/module.dart';
import 'abs.dart';
import 'ceil.dart';
import 'floor.dart';
import 'max.dart';
import 'min.dart';
import 'round.dart';
import 'trunc.dart';

/// Utility module - mathematical utility functions
/// Mirrors Haskell Glue.Lib.Math.Utility exactly

/// The utility module containing mathematical utility functions
/// Mirrors Haskell Glue.Lib.Math.Utility.utility exactly
final ModuleInfo utility = nativeModule('ffi.math.utility', [
  // Absolute value
  ('abs', IrNativeFunc(abs)),

  // Rounding functions (always return Integer)
  ('floor', IrNativeFunc(floor)),
  ('ceil', IrNativeFunc(ceil)),
  ('round', IrNativeFunc(round)),
  ('trunc', IrNativeFunc(trunc)),

  // Min/max functions (two arguments)
  ('min', IrNativeFunc(min)),
  ('max', IrNativeFunc(max)),
]);

File: haskell/glue/src/Glue/Lib/Math/Utility.hs
module Glue.Lib.Math.Utility where

import Glue.Eval (Eval)
import Glue.IR (IR (..))
import Glue.Lib.Math.Utility.Abs qualified as Abs
import Glue.Lib.Math.Utility.Ceil qualified as Ceil
import Glue.Lib.Math.Utility.Floor qualified as Floor

import Glue.Lib.Math.Utility.Max qualified as Max
import Glue.Lib.Math.Utility.Min qualified as Min
import Glue.Lib.Math.Utility.Round qualified as Round
import Glue.Lib.Math.Utility.Trunc qualified as Trunc
import Glue.Module (ModuleInfo, nativeModule)

utility :: ModuleInfo Eval
utility =
    nativeModule
        "ffi.math.utility"
        [ ("abs", NativeFunc Abs.abs)
        , ("floor", NativeFunc Floor.floor)
        , ("ceil", NativeFunc Ceil.ceil)
        , ("round", NativeFunc Round.round)
        , ("trunc", NativeFunc Trunc.trunc)
        , ("min", NativeFunc Min.min)
        , ("max", NativeFunc Max.max)
        ]
